lib/api/objectPutCopyPart.js

```diff
@@ -0,0 +1,356 @@
+            return services.metadataValidateAuthorization(valGetParams,
+                (err, sourceBucketMD, sourceObjMD) => {
+                    if (err) {
+                        log.debug('error validating get part of request',
+                        { error: err });
+                        return next(err);
+                    }
+                    if (!sourceObjMD) {
+                        log.debug('no source object', { sourceObject });
+                        return next(errors.NoSuchKey);
+                    }
+                    const headerValResult =
+                        validateHeaders(sourceObjMD, request.headers);
+                    if (headerValResult.error) {
+                        return next(errors.PreconditionFailed);
+                    }
+                    let dataLocator;
+                    // If 0 byte object just set dataLocator to empty array
+                    if (!sourceObjMD.location) {
+                        dataLocator = [];
+                    } else {
+                        // To provide for backwards compatibility before
+                        // md-model-version 2, need to handle cases where
+                        // objMD.location is just a string
+                        dataLocator = Array.isArray(sourceObjMD.location) ?
+                        sourceObjMD.location : [{ key: sourceObjMD.location }];
+                    }
+
+                    if (sourceObjMD['x-amz-server-side-encryption']) {
+                        for (let i = 0; i < dataLocator.length; i++) {
+                            dataLocator[i].masterKeyId = sourceObjMD
+                            ['x-amz-server-side-encryption-aws-kms-key-id'];
+                            dataLocator[i].algorithm =
+                                sourceObjMD['x-amz-server-side-encryption'];
+                        }
+                    }
+
+                    const sourceSize =
+                        parseInt(sourceObjMD['content-length'], 10);
+                    // If part size is greater than 5GB, reject it
+                    if (sourceSize > 5368709120) {
+                        return callback(errors.EntityTooLarge);
+                    }
+                    let copyObjectSize = sourceSize;
+                    if (request.headers['x-amz-copy-source-range']) {
+                        const range = parseRange(request
+                            .headers['x-amz-copy-source-range'],
+                            sourceSize);
+                        // If have a data model before version 2, cannot
+                        // support get range copy (do not have size
+                        // stored with data locations)
+                        if ((range && dataLocator.length >= 1) &&
+                            (dataLocator[0].start === undefined
+                            || dataLocator[0].size === undefined)) {
+                            return callback(errors.NotImplemented);
```

Comment:
We should introduce an Arsenal error for this case, something like `errors.LegacyObject`. We could lump all incompatible actions due to breaking changes into this error with a custom error description. Also we would need a log here for debugging.